datafusion/functions/src/macros.rs

```diff
@@ -0,0 +1,123 @@
+// under the License.
+
+/// macro that exports a list of function names as:
+/// 1. individual functions in an `expr_fn` module
+/// 2. a single function that returns a list of all functions
+///
+/// Equivalent to
+/// ```text
+/// pub mod expr_fn {
+///     use super::*;
+///     /// Return encode(arg)
+///     pub fn encode(args: Vec<Expr>) -> Expr {
+///         super::encode().call(args)
+///     }
+///  ...
+/// /// Return a list of all functions in this package
+/// pub(crate) fn functions() -> Vec<Arc<ScalarUDF>> {
+///     vec![
+///       encode(),
+///       decode()
+///    ]
+/// }
+/// ```
+macro_rules! export_functions {
+    ($(($FUNC:ident,  $($arg:ident)*, $DOC:expr)),*) => {
+        pub mod expr_fn {
+            $(
+                #[doc = $DOC]
+                /// Return $name(arg)
+                pub fn $FUNC($($arg: datafusion_expr::Expr),*) -> datafusion_expr::Expr {
+                    super::$FUNC().call(vec![$($arg),*],)
+                }
+            )*
+        }
+
+        /// Return a list of all functions in this package
+        pub fn functions() -> Vec<std::sync::Arc<datafusion_expr::ScalarUDF>> {
+            vec![
+                $(
+                    $FUNC(),
+                )*
+            ]
+        }
+    };
+}
+
+/// Creates a singleton `ScalarUDF` of the `$UDF` function named `$GNAME` and a
+/// function named `$NAME` which returns that function named $NAME.
+///
+/// This is used to ensure creating the list of `ScalarUDF` only happens once.
+macro_rules! make_udf_function {
+    ($UDF:ty, $GNAME:ident, $NAME:ident) => {
+        /// Singleton instance of the function
+        static $GNAME: std::sync::OnceLock<std::sync::Arc<datafusion_expr::ScalarUDF>> =
+            std::sync::OnceLock::new();
+
+        /// Return a [`ScalarUDF`] for [`$UDF`]
+        ///
+        /// [`ScalarUDF`]: datafusion_expr::ScalarUDF
+        fn $NAME() -> std::sync::Arc<datafusion_expr::ScalarUDF> {
+            $GNAME
+                .get_or_init(|| {
+                    std::sync::Arc::new(datafusion_expr::ScalarUDF::new_from_impl(
+                        <$UDF>::new(),
+                    ))
+                })
+                .clone()
+        }
+    };
+}
+
+/// Macro creates the named module if the feature is enabled
+/// otherwise creates a stub
+///
+/// Which returns:
+///
+/// 1. The list of actual function implementation when the relevant
+/// feature is activated,
+///
+/// 2. A list of stub function when the feature is not activated that produce
+/// a runtime error (and explain what feature flag is needed to activate them).
+///
+/// The rationale for providing stub functions is to help users to configure datafusion
+/// properly (so they get an error telling them why a function is not available)
+/// instead of getting a cryptic "no function found" message at runtime.
+
+macro_rules! make_package {
+    ($name:ident, $feature:literal, $DOC:expr) => {
+        #[cfg(feature = $feature)]
```

Comment:
Hmm, I wonder if all function packages have corresponding features? I guess for built-in scalar functions they don't have feature?

Reply:
This is an excellent observation

I think eventually it would be nice to have as many functions as possible be optional and behind a feature flat. There are some functions that back built in `Expr` like `make_array`   

What I am thinking is that we'll first break out the functions that do have their own feature flags already and establish a pattern and then we can design / lift out others that don't. I hope to write down more of the plan shortly

Reply:
Here is a discussion / proposal on how the functions could be organized: https://github.com/apache/arrow-datafusion/issues/9100